### PR TITLE
🚨All attention refactor🚨

### PR DESCRIPTION
**Pull request by ArthurZucker**
# What does this PR do?
Todo in this PR:

- [ ] Cohere
- [ ] Chameleon
- [ ] DBRX
- [x] Gemma
- [x] Gemma2
- [x] GLM (modular donc rien à faire je crois)
- [ ] gpt_neoX et GPT2
- [x] Granite
- [ ] Jamba
- [ ] JetMoe
- [ ] Mimi
- [x] Mistral
- [x] Mixtral
- [ ] Mllama
- [ ] Moshi
- [ ] Nemotron
- [ ] OPT
- [x] Phi
- [ ] Ph3
- [ ] PhiMoe
- [x] Qwen2
- [ ] qwen2Moe
- [ ] qwen2VL
- [ ] SableML
- [x] StartCoder2 -> Modular normalement oK
- [ ] Idefics1,2,3
- [x] Olmo
- [x] Olmo2
- [ ] Siglip
- [ ] Whisper

### CODE REVIEW

File: src/transformers/modeling_utils.py
Comment (ArthurZucker) [inline src/transformers/modeling_utils.py:5577]: This should help with kwargs as well

**Review by ArthurZucker — COMMENTED** [review]
super small nit

File: src/transformers/models/decision_transformer/modeling_decision_transformer.py
Comment (ArthurZucker) [reply src/transformers/models/decision_transformer/modeling_decision_transformer.py:104]: It might be what's failing your tests!

File: src/transformers/models/pixtral/modeling_pixtral.py
Comment (ArthurZucker) [inline src/transformers/models/pixtral/modeling_pixtral.py:216]: it's now the. same as llama now

File: src/transformers/models/qwen2/modular_qwen2.py
Comment (ArthurZucker) [inline src/transformers/models/qwen2/modular_qwen2.py:76]: let's just add a comment for main diff with llama!

File: src/transformers/models/qwen2_moe/modeling_qwen2_moe.py
Comment (ArthurZucker) [inline src/transformers/models/qwen2_moe/modeling_qwen2_moe.py:1563]: this one's breaking no? To revert!

File: src/transformers/models/starcoder2/modular_starcoder2.py
Comment (ArthurZucker) [inline src/transformers/models/starcoder2/modular_starcoder2.py:54]:
```suggestion
```

File: src/transformers/models/starcoder2/modular_starcoder2.py
Comment (ArthurZucker) [inline src/transformers/models/starcoder2/modular_starcoder2.py:81]: init can be the same as llama as well I think we can standardize more!

File: src/transformers/models/starcoder2/modular_starcoder2.py
Comment (ArthurZucker) [inline src/transformers/models/starcoder2/modular_starcoder2.py:138]:
```suggestion
        attn_output = nn.functional.dropout(attn_output, p=self.residual_dropout, training=self.training) # main diff with llama
```

File: src/transformers/models/starcoder2/modular_starcoder2.py
Comment (ArthurZucker) [inline src/transformers/models/starcoder2/modular_starcoder2.py:153]: I dont'think we should be needing this!

File: src/transformers/models/zamba/modeling_zamba.py
Comment (ArthurZucker) [inline src/transformers/models/zamba/modeling_zamba.py:773]: same as llama now! More standards!

File: tests/models/encoder_decoder/test_modeling_encoder_decoder.py
Comment (ArthurZucker) [inline tests/models/encoder_decoder/test_modeling_encoder_decoder.py:736]: perfect

File: tests/models/gpt2/test_modeling_gpt2.py
Comment (ArthurZucker) [inline tests/models/gpt2/test_modeling_gpt2.py:510]: that might be an issue, we can ping @michaelbenayoun maybe!

**Review by ArthurZucker — COMMENTED** [review]
Cannot approve as it's mine hehe but APPROVED 🟢

File: src/transformers/models/gemma2/modular_gemma2.py
Comment (Cyrilvallez) [reply src/transformers/models/gemma2/modular_gemma2.py:268]: I agree, I dislike it as well...

File: src/transformers/models/mixtral/modular_mixtral.py
Comment (Cyrilvallez) [reply src/transformers/models/mixtral/modular_mixtral.py:316]: In the output as well, otherwise we could have used `super().forward()` 🥲

File: src/transformers/models/pixtral/modeling_pixtral.py
Comment (Cyrilvallez) [reply src/transformers/models/pixtral/modeling_pixtral.py:216]: nop - biases are False!

File: src/transformers/models/olmoe/modeling_olmoe.py
Comment (Cyrilvallez) [reply src/transformers/models/olmoe/modeling_olmoe.py:273]: nop - biases are False!

File: src/transformers/models/qwen2_moe/modeling_qwen2_moe.py
Comment (Cyrilvallez) [reply src/transformers/models/qwen2_moe/modeling_qwen2_moe.py:1563]: Indeed! And in Mistral!

File: src/transformers/models/starcoder2/modular_starcoder2.py
Comment (Cyrilvallez) [reply src/transformers/models/starcoder2/modular_starcoder2.py:81]: nop - the bias term in the config is different 🥲

**Review by LysandreJik — APPROVED** [review]
Very impressive work, kudos to you both!

**Comment by Cyrilvallez** [comment]
Confirmed slow tests with Llama, everything is similar to main!

**Comment by ydshieh** [comment]
run-slow: vit

(just a check unrelated to this PR)

File: src/transformers/integrations/flash_attention.py
Comment (vasqu) [inline src/transformers/integrations/flash_attention.py:35]: Any reason we specifically force float16? This check has been moved to the fa utils when providing a target dtype - seems like we have doubled semantics, imo it should be either but not both.

Comment (vasqu) [reply src/transformers/integrations/flash_attention.py:35]: Are all rope implementations fp32 agnostic? Iirc there has been a recast of qk to v dtype.

File: src/transformers/integrations/flex_attention.py
Comment (vasqu) [inline src/transformers/integrations/flex_attention.py:8]: There is `is_torch_flex_attn_available()` in the utils. Might be more general and easier to change especially if conditions change on when flex attn is allowed.

File: src/transformers/integrations/flex_attention.py
Comment (vasqu) [inline src/transformers/integrations/flex_attention.py:20]: Head mask could be added as well as done in gpt neox.

File: src/transformers/integrations/sdpa_attention.py
Comment (vasqu) [inline src/transformers/integrations/sdpa_attention.py:39]: Would love to have comments back on why we call contiguous seems out of the blue otherwise.

File: src/transformers/integrations/sdpa_attention.py
Comment (vasqu) [inline src/transformers/integrations/sdpa_attention.py:42]: Same here for comment

File: src/transformers/integrations/flash_attention.py
Comment (vasqu) [inline src/transformers/integrations/flash_attention.py:44]: Would love to see kwarg usage to identify what we pass more explicitly.

File: src/transformers/integrations/flex_attention.py
Comment (vasqu) [inline src/transformers/integrations/flex_attention.py:40]: Same for let's add comments back.

**Review by vasqu — COMMENTED** [review]
Great work guys! I think there might be value in keeping some comments, e.g. why call contiguous on sdpa, and clarifying the fa usage on recasting to half (which originates from PEFT and/or rope).

File: src/transformers/integrations/flash_attention.py
Comment (ArthurZucker) [reply src/transformers/integrations/flash_attention.py:35]: I think this is more a typo and as you say we gotta default to flash attention's call!

File: src/transformers/integrations/flex_attention.py
Comment (ArthurZucker) [reply src/transformers/integrations/flex_attention.py:20]: It's more like gpt_neo_x will get a flex attention function that has support for head mask! But good point!

File: src/transformers/integrations/sdpa_attention.py
Comment (ArthurZucker) [reply src/transformers/integrations/sdpa_attention.py:39]: cc @Cyrilvallez

**Comment by SimJeg** [comment]
@ArthurZucker @Cyrilvallez do you plan to refactor `modeling_phi3.py` using the `ALL_ATTENTION_FUNCTIONS` you recently introduced ? I see Phi 3 is in the list shared by @ArthurZucker at the beginning of this PR. It would be very helpful for our [kvpress](https://github.com/NVIDIA/kvpress) package. We plan to update it to be compatible with the future v4.48, any idea of when it will be released ? (december ? january ?)

**Comment by ArthurZucker** [comment]
Release will happen on 🎅🏻 🎁 !
Yeah for sure. Unless you submit a PR first 👀 not sure it will be in this release as we are all going on holidays but in january's release it will be included

**Comment by SimJeg** [comment]
I'm going to holidays too so I'll wait for January ^^ Happy holidays!

**Comment by ArthurZucker** [comment]
Happy holidays!

**Comment by poedator** [comment]
I noticed that there is still torch.reshape used in quite a few places, for instance in `modeling_gpt2.py`  Won't it be an obstacle to compiling these models? Why not replacing it with einops and using `einops._torch_specific.allow_ops_in_compiled_graph()`  ?

**Comment by Cyrilvallez** [comment]
Hey! Indeed for gpt2 I used a 'reshape' instead of the usual 'view' because I hit an edge case that wasn't compatible with viewing at some point (but I will recheck that it still appears with latest developments, might have been an artifact during the debugging process). Whenever possible (most of the time), 'reshape' is actually equivalent to 'view' so no worries there anyway 😉

**Comment by ArthurZucker** [comment]
Einops has never been used in our modeling codes, and we avoid using it at all costs! `Llama` uses `reshape` and pretty sur it's compile compatible! 🤗

**Comment by ArthurZucker** [comment]
Yeah sorry @BenjaminBossan, it's just that we don't use that attribute so not sure we are gonna add it back! THe PR is breaking as the name indicates! But I hope it was not too much trouble!

**Comment by BenjaminBossan** [comment]
Thanks for letting me know, it should be an easy fix on the PEFT side.

**Comment by foreverpiano** [comment]
query_states = query_states.view(bsz, q_len, self.num_heads, self.head_dim).transpose(1, 2)

    raise AttributeError(f"'{type(self).__name__}' object has no attribute '{name}'")
AttributeError: 'MistralAttention' object has no attribute 'num_heads'

How can I fix this?

**Comment by ArthurZucker** [comment]
Hey! you should try to use the latest release of `transformers`! `query_states = self.q_proj(hidden_states).view(hidden_shape).transpose(1, 2)` is what's used now.

**Comment by foreverpiano** [comment]
Have you tested on several benchmarks about the performance? I knew that the [Longbench](https://github.com/THUDM/LongBench) score on transformer v4.47 vs v4.36 varies a lot on llama-3. Is it stable on this version?

**Comment by Cyrilvallez** [comment]
Hey! Everything stays the same in terms of user experience/benchmark scores. If you used to hack into the different Layer classes however, it may have changed a bit. You can simply go and check-out the modeling code in this case (as was the case if you hacked into it in the first place I guess!)

**Comment by poedator** [comment]
My friends use a `GPT2Model` in production and want to compile it with `StaticCache`. With the maintainers blessing, I would try to create a PR with `DynamicCache` / `StaticCache` support in GPT2Model. 
I am quite familiar with `Cache` class, I already coded some and made the `DynamicCache` work.

Please let me know if there are any hidden obstacles in Cache implementation for GPT2? Which tests to run or add? 
@ArthurZucker

**Comment by Rocketknight1** [comment]
cc @gante to that question!

**Comment by gante** [comment]
I've chatted to @poedator offline -- I couldn't think of any obstacle in particular, and suggested a) to ensure we leave a deprecation warning regarding the old cache format b) use `RUN_SLOW=1 py.test tests/models/gpt2/test_modeling_gpt2.py` as a correctness check (gpt2 is fairly well tested, especially wrt text generation)

**Comment by poedator** [comment]
It looks like `test_flash_attn_2_from_config` is broken - it expects attention layer to have flashattention in its name, 
` if "FlashAttention" in module.__class__.__name__:...`
but after this refactoring, the attention classes are named differently. 

ref https://github.com/huggingface/transformers/blob/5fa35344755d8d9c29610b57d175efd03776ae9e/tests/test_modeling_common.py#L4641

please fix or suspend the test.
@ArthurZucker

**Comment by ArthurZucker** [comment]
indeed gimme a min!

File: src/transformers/integrations/sdpa_attention.py
Comment (yuanyao-nv) [inline src/transformers/integrations/sdpa_attention.py:30]: @ArthurZucker Wondering if you know whether this repeat_kv is needed here? `torch.nn.functional.scaled_dot_product_attention` allows num heads to be different for q and k/v, and having this expansion just defeats the purpose of doing GQA.

Comment (vasqu) [reply src/transformers/integrations/sdpa_attention.py:30]: @yuanyao-nv Sadly, yes it is due to multiple reasons:
- GQA support was only later added to sdpa (we support up to torch 2.1)
- The GQA support in sdpa is very limited
  - Quoting https://docs.pytorch.org/docs/stable/generated/torch.nn.functional.scaled_dot_product_attention.html "It currently works only for Flash_attention and math kernel on CUDA tensor"
  - However, since we more often than not use a mask, we require the memory-efficient backend (fa cannot work with masks in sdpa) to avoid the least efficient math backend

Comment (ArthurZucker) [reply src/transformers/integrations/sdpa_attention.py:30]: On point!

Comment (yuanyao-nv) [reply src/transformers/integrations/sdpa_attention.py:30]: Thanks for the reply!
With regard to the flash attention support gap, how does introducing repeat_kv solve the issue? Does flash attention have better support for masked cases in non-GQA than GQA?
The reason I'm asking is that I'm trying to export models into FX graph and ONNX and I'm trying to preserve the SDPA op for Attention in the FX graph. The repeat_kv introduces additional ops that can lead to inefficiencies when a non-torch backend tries to parse such graphs.
I'm wondering if the use of repeat_kv should be configurable based on which backend is intended?

Comment (yuanyao-nv) [reply src/transformers/integrations/sdpa_attention.py:30]: @vasqu @ArthurZucker any thoughts on the above questions? Thanks.

Comment (vasqu) [reply src/transformers/integrations/sdpa_attention.py:30]: Hey @yuanyao-nv,

sorry for the late reply. The original flash attention (not as backend for sdpa) allows proper GQA support - usually when we have a mask, we need to opt for the varlen implementation then (otherwise we take the "normal" function). 

"I'm wondering if the use of repeat_kv should be configurable based on which backend is intended?" - Not sure about that but also not 100% against it. My concern is that we would need to detect whether we use masks and what backends are enabled by the user; in most cases I would expect batches and some sort of padding.

Comment (yuanyao-nv) [reply src/transformers/integrations/sdpa_attention.py:30]: Another followup question (might be a naive one): 
The use case you're referring to above seems to be `attn_implementation="sdpa"` and also explicitly requesting the FA backend (maybe via a context manager `with sdpa_kernel(backends=[SDPBackend.FLASH_ATTENTION]):`?) 
However, if the user really intends to use FA, shouldn't they directly set `attn_imlementation="flash_attention_2"` or `"flash_attention_3"`? ie, why would they set `attn_implementation="sdpa"` and then request the FA kernel?

Comment (ArthurZucker) [reply src/transformers/integrations/sdpa_attention.py:30]: Yes, in general using `flash_attention_2` is just the best / recommended way!

Comment (yuanyao-nv) [reply src/transformers/integrations/sdpa_attention.py:30]: Thanks for confirming. In that case is there anything stopping us from removing `repeat_kv` from the sdpa_attention path? It sounds like if the user really wants to use the FA backend they should not need to use this path in the first place.
And this will make the exported fx graph and ONNX graph more efficient and easier for other backends to work with.
I'd be happy to submit a PR for this.

Comment (vasqu) [reply src/transformers/integrations/sdpa_attention.py:30]: I think https://github.com/huggingface/transformers/pull/35235#discussion_r2176860497 still explains most reasons:
- `enable_gqa` is not available for earlier versions of torch (starts with 2.5.x)
- If we have a mask, we need to avoid the math kernel (glad to see benchmarks to prove me wrong there) and instead use the xformers one (memory-efficient backend)
  - `enable_gqa` works only with fa or math
  - If we have a mask, it already cuts out the fa kernel (torch internal restriction)
  - Fallback to math kernel
- It's not about users requesting specific backends but moreso to avoid entering inefficient branches (math kernel) - if a user uses SDPA they should expect the more efficient backends...

The original flash attention is better in that way but SDPA is the standard attention as it's native torch.

Comment (vasqu) [reply src/transformers/integrations/sdpa_attention.py:30]: FYI, there is #39412 which partially enables this kwarg

**Pull request merged**
